File: bk2ch23p818xml/ch36p1070xml/MyXMLParserDelegate.m


#import "MyXMLParserDelegate.h"


@implementation MyXMLParserDelegate

- (void) start: (NSString*) el parent: (id) p {
    self.name = el;
    self.parent = p;
    self.text = [NSMutableString string];
}

- (void) makeChild: (Class) class elementName: (NSString*) elementName parser: (NSXMLParser*) parser {
    MyXMLParserDelegate* del = [class new];
    self.child = del;
    parser.delegate = del;
    [del start: elementName parent: self];
}

- (void) finishedChild: (NSString*) s { // subclass implements as desired
}

- (void)parser:(NSXMLParser *)parser foundCharacters:(NSString *)string {
    [self.text appendString:string];
}

- (void)parser:(NSXMLParser *)parser didEndElement:(NSString *)elementName namespaceURI:(NSString *)namespaceURI qualifiedName:(NSString *)qName {
    if (self.parent) {
        [self.parent finishedChild: [self.text copy]];
        parser.delegate = self.parent;
    }
}



@end

File: bk2ch23p818xml/ch36p1070xml/MyXMLParserDelegate.swift


import UIKit

class MyXMLParserDelegate : NSObject, NSXMLParserDelegate {
    var text : String!
    var name : String!
    weak var parent : MyXMLParserDelegate?
    var child : MyXMLParserDelegate!
    
    required override init() {
        super.init()
    }
    
    func start(el:String, parent:MyXMLParserDelegate) {
        self.name = el
        self.parent = parent
        if self.text == nil {
            self.text = ""
        }
    }
    
    func makeChild(klass:MyXMLParserDelegate.Type, elementName:String, parser:NSXMLParser) {
        let del = klass()
        self.child = del
        parser.delegate = del
        del.start(elementName, parent:self)
    }
    
    func finishedChild(s:String) {} // subclass must implement as desired
    
    // NSXMLParser delegate messages
    
    func parser(parser: NSXMLParser, foundCharacters string: String!) {
        if self.text == nil {
            return
        }
        self.text = self.text + string
    }
    
    func parser(parser: NSXMLParser, didEndElement elementName: String!, namespaceURI: String!, qualifiedName qName: String!) {
        if self.parent != nil {
            self.parent!.finishedChild(self.text)
            parser.delegate = self.parent
        }
    }
    
}
